File: skills/test-graph/test_graph/sources/DeployCluster.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//SOURCES ../sdk/java/src/main/java/com/hayden/testgraphsdk/sdk/*.java

import java.util.List;
import java.util.Map;

import com.hayden.testgraphsdk.sdk.ClusterLifecycle;
import com.hayden.testgraphsdk.sdk.ClusterLifecyclePlan;
import com.hayden.testgraphsdk.sdk.Node;
import com.hayden.testgraphsdk.sdk.NodeResult;
import com.hayden.testgraphsdk.sdk.NodeSpec;

public class DeployCluster {
    private record Target(String target, String backend, String dispatch, boolean explicit) {}

    private static final List<Target> TARGETS = List.of(
            new Target("local-preview", "local", "local", false),
            new Target("local-github-action", "github-action", "github-action", false),
            new Target("aws-preview", "aws", "aws", true)
    );

    private static final NodeSpec SPEC = NodeSpec.of("tg6.lifecycle.java.deploy-cluster")
            .kind(NodeSpec.Kind.ASSERTION)
            .tags("tg6", "environment", "lifecycle-template", "jbang")
            .output("Runtime", "string")
            .output("DeployCases", "string");

    public static void main(String[] args) {
        Node.run(args, SPEC, ctx -> {
            int cases = 0;
            boolean dispatchMatches = true;
            boolean missingProvisions = true;
            boolean existingReuses = true;
            for (Target target : TARGETS) {
                ClusterLifecyclePlan missing = ClusterLifecycle.deployCluster(target.target(), target.backend(), false);
                ClusterLifecyclePlan existing = ClusterLifecycle.deployCluster(target.target(), target.backend(), true);
                cases += 2;
                dispatchMatches &= missing.dispatchKey().equals(target.dispatch());
                missingProvisions &= missing.expectedState().equals("provision-missing");
                existingReuses &= existing.expectedState().equals("reuse-existing");
            }

            ClusterLifecyclePlan sample = ClusterLifecycle.deployCluster("local-preview", "local", false);
            NodeResult result = NodeResult.pass(ctx.nodeId())
                    .assertion("all_targets_covered", cases == TARGETS.size() * 2)
                    .assertion("deploy_uses_provision_action", sample.environmentAction().equals("provision"))
                    .assertion("deploy_uses_apply_command", sample.tofuCommand().equals(List.of("tofu", "apply", "-auto-approve")))
                    .assertion("missing_environment_is_provisioned", missingProvisions)
                    .assertion("existing_environment_is_reused", existingReuses)
                    .assertion("dispatch_metadata_matches_target_matrix", dispatchMatches)
                    .assertion("aws_requires_explicit_selection", ClusterLifecycle.deployCluster("aws-preview", "aws", false).requiresExplicitSelection())
                    .publish("Runtime", "jbang")
                    .publish("DeployCases", Integer.toString(cases));
            for (Map.Entry<String, String> entry : sample.published().entrySet()) {
                result.publish("Deploy" + entry.getKey(), entry.getValue());
            }
            return result;
        });
    }
}

File: skills/test-graph/test_graph/sources/deploy_cluster.py
# /// script
# requires-python = ">=3.10"
# dependencies = ["testgraphsdk"]
#
# [tool.uv.sources]
# testgraphsdk = { path = "../sdk/python", editable = true }
# ///
from __future__ import annotations

from testgraphsdk import NodeResult, NodeSpec, deploy_cluster, node


TARGETS = (
    ("local-preview", "local", "local", False),
    ("local-github-action", "github-action", "github-action", False),
    ("aws-preview", "aws", "aws", True),
)

SPEC = (
    NodeSpec("tg6.lifecycle.python.deploy-cluster")
    .kind("assertion")
    .tags("tg6", "environment", "lifecycle-template", "uv")
    .output("Runtime")
    .output("DeployCases")
)


@node(SPEC)
def main(ctx):
    plans = [
        deploy_cluster(target, backend, environment_exists=environment_exists)
        for target, backend, _dispatch, _requires_explicit in TARGETS
        for environment_exists in (False, True)
    ]

    result = (
        NodeResult.pass_(ctx.node_id)
        .assertion("all_targets_covered", len(plans) == len(TARGETS) * 2)
        .assertion("deploy_uses_provision_action", all(plan.environment_action == "provision" for plan in plans))
        .assertion("deploy_uses_apply_command", all(plan.tofu_command == ("tofu", "apply", "-auto-approve") for plan in plans))
        .assertion("missing_environment_is_provisioned", all(
            deploy_cluster(target, backend, environment_exists=False).expected_state == "provision-missing"
            for target, backend, _dispatch, _requires_explicit in TARGETS
        ))
        .assertion("existing_environment_is_reused", all(
            deploy_cluster(target, backend, environment_exists=True).expected_state == "reuse-existing"
            for target, backend, _dispatch, _requires_explicit in TARGETS
        ))
        .assertion("dispatch_metadata_matches_target_matrix", all(
            deploy_cluster(target, backend).dispatch_key == dispatch
            for target, backend, dispatch, _requires_explicit in TARGETS
        ))
        .assertion("aws_requires_explicit_selection", deploy_cluster("aws-preview", "aws").requires_explicit_selection)
        .publish("Runtime", "uv")
        .publish("DeployCases", str(len(plans)))
    )
    for key, value in deploy_cluster("local-preview", "local").published().items():
        result.publish(f"Deploy{key}", value)
    return result


if __name__ == "__main__":
    main()
